lib/models/addon.js

```diff
@@ -215,12 +221,26 @@ let addonProto = {
 
     this.treeForMethods = defaultsDeep({}, DEFAULT_TREE_FOR_METHODS);
 
-    p.setupRegistry(this);
+    if (this.parent) {
+      // The parent should already have a _packageInfoCache.
+      // Because it does, this package should already be in the cache.
+      this._packageInfoCache = this.parent._packageInfoCache;
+    } else {
+      // this is a 'root' addon, so create a new PackageInfoCache
+      // for it so we get the PackageInfo for it and its children.
+      this._packageInfoCache = new PackageInfoCache(this.ui);
+    }
 
-    if (!this.name) {
-      throw new SilentError(`An addon must define a \`name\` property (found at ${this.root}).`);
+    this._packageInfo = this._packageInfoCache.loadAddon(this);
+
+    // force us to use the real path as the root.
+    if (this.root !== this._packageInfo.realPath) {
```

Comment:
I removed the test, but would like to keep the determining of real paths in PIC, instead of here. It's already set up there to do it (using getRealDirectoryPath), and does set the value back here in Addon. Same as project.